src/category_theory/instances/Top/stalks.lean

```diff
@@ -0,0 +1,79 @@
+-- Copyright (c) 2019 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison
+
+import category_theory.instances.Top.open_nhds
+import category_theory.instances.Top.presheaf
+import category_theory.limits.limits
+
+universes v u v' u'
+
+open category_theory
+open category_theory.instances
+open category_theory.instances.Top
+open category_theory.limits
+open topological_space
+
+variables {C : Type u} [𝒞 : category.{v+1} C]
+include 𝒞
+
+variables [has_colimits.{v} C]
+
+variables {X Y Z : Top.{v}}
+
+namespace category_theory.instances.Top.presheaf
+
+variables (C)
+/-- Stalks are functorial with respect to morphisms of presheaves over a fixed `X`. -/
+def stalk_functor (x : X) : X.presheaf C ⥤ C :=
+((whiskering_left _ _ C).obj (open_nhds.inclusion x).op) ⋙ colim
+
+variables {C}
+
+/--
+The stalk of a presheaf `F` at a point `x` is calculated as the colimit of the functor
+nbhds x ⥤ opens F.X ⥤ C
+-/
+def stalk (ℱ : X.presheaf C) (x : X) : C :=
+(stalk_functor C x).obj ℱ -- -- colimit (nbhds_inclusion x ⋙ ℱ)
+
+@[simp] lemma stalk_functor_obj (ℱ : X.presheaf C) (x : X) : (stalk_functor C x).obj ℱ = ℱ.stalk x := rfl
+
+variables (C)
+
+def stalk_pushforward (f : X ⟶ Y) (ℱ : X.presheaf C) (x : X) : (f _* ℱ).stalk (f x) ⟶ ℱ.stalk x :=
+begin
+  -- This is a hack; Lean doesn't like to elaborate the term written directly.
```

Comment:
@digama0 Do you have any idea what might be going on here? It feels like a bad sign to me.

Reply:
To type-check this definition, you have to apply `(F ⋙ G) ⋙ H ≅ F ⋙ (G ⋙ H)` and `(F ⋙ G).op ≅ F.op ⋙ G.op`, both of which are true by reflexivity.

The left endpoint of the first natural transformation is
```
colim.obj ((open_nhds.inclusion (⇑f x) ⋙ opens.map f).op ⋙ ℱ)
```
while the expected left endpoint is (after unfolding some definitions)
```
colim.obj ((open_nhds.inclusion (⇑f x)).op ⋙ ((opens.map f).op ⋙ ℱ))
```
So it makes sense (to me) that Lean has some difficulties parsing a definition without extra type ascriptions, If I explicitly insert the associator
```
def stalk_pushforward (f : X ⟶ Y) (ℱ : X.presheaf C) (x : X) : (f _* ℱ).stalk (f x) ⟶ ℱ.stalk x :=
colim.map ((functor.associator _ _ _).inv ≫
  whisker_right (nat_trans.op (open_nhds.inclusion_map_iso f x).inv) ℱ) ≫
colimit.pre ((open_nhds.inclusion x).op ⋙ ℱ) (open_nhds.map f x).op
```
or I add some type ascription
```
def stalk_pushforward (f : X ⟶ Y) (ℱ : X.presheaf C) (x : X) : (f _* ℱ).stalk (f x) ⟶ ℱ.stalk x :=
(colim.map (whisker_right (nat_trans.op (open_nhds.inclusion_map_iso f x).inv) ℱ) :
  colim.obj ((open_nhds.inclusion (f x) ⋙ opens.map f).op ⋙ ℱ) ⟶ _) ≫
colimit.pre ((open_nhds.inclusion x).op ⋙ ℱ) (open_nhds.map f x).op
```
then Lean accepts the definition. In these examples I still silently use the following natural iso:
```
def op_comp (F : C ⥤ D) (G : D ⥤ E) : (F ⋙ G).op ≅ F.op ⋙ G.op := iso.refl _
```
This one is less dangerous to do silently than associativity. With associativity, if Lean needs to unify
`(F ⋙ G) ⋙ H =?= F' ⋙ (G' ⋙ H')` during elaboration, then it will try to match `F ⋙ G =?= F'` and `H =?= G' ⋙ H'` which will go wrong. With `op_comp` the outermost operation is different, so it cannot make progress unless it unfolds the definition, in which case it can match `(F ⋙ G).op =?= F'.op ⋙ G'.op` (if `F` unifies with `F'` and `G` with `G'`).

Reply:
Thanks, @fpvandoorn, that is a very helpful explanation!

Reply:
I've incorporated your "explicit associator" solution.

Reply:
... and then realised I don't know how to fix the subsequent proofs. This still needs work.